lib/gameboard/board.rb

```diff
@@ -0,0 +1,495 @@
+    #
+    def delta(point, slope, coords = false)
+      delta =  point.zip(slope).map {|point| point.reduce(:+) }
+      piece = board.find { |cell| cell.coord.position == delta }
+      raise "Off Grid" unless !!piece
+      coords ? piece.coord.position : piece.value
+    end
+
+    # Public: Return every diagonal on the gameboard. It does not require a
+    #         square board to function.
+    #
+    # coords: An optional boolean to specify returning an Array of coordinates
+    #         instead of values.
+    #
+    # Example
+    #      -------------
+    #    0 | 1 | 2 | 3 |
+    #      -------------
+    #    1 | A | B | C |
+    #      -------------
+    #    2 | X | Y | Z |
+    #      -------------
+    #      - 0 - 1 - 2 -
+    #
+    #
+    #   diagonal
+    #     #=> [
+    #          ["X", "B", 3], [1, "B", "Z"],
+    #          ["A", 2], ["A", "Y"], [1], ["X"],
+    #          ["Y", "C"], [2, "C"], ["Z"], [3]
+    #         ]
+    #
+    #   diagonal(true)
+    #     #=> [
+    #          [[0, 0], [1, 1], [2, 2]],
+    #          [[0, 2], [1, 1], [2, 0]],
+    #          [[0, 1], [1, 2]],
+    #          [[0, 1], [1, 0]],
+    #          [[0, 2]],
+    #          [[0, 0]],
+    #          [[1, 0], [2, 1]],
+    #          [[1, 2], [2, 1]],
+    #          [[2, 0]],
+    #          [[2, 2]]
+    #         ]
+    #
+    def diagonal(coords = false)
+      diagonals = []
+
+      height.times do |i|
+        diagonals << get_diagonal([0, i], coords)
+        diagonals << get_diagonal([0, height - 1 - i], coords, false)
+      end
+      (1...width).each do
+        |i| diagonals << get_diagonal([i, 0], coords)
+        diagonals << get_diagonal([i, height - 1], coords, false)
+      end
+
+      diagonals
+    end
+
+    # Public: Find a cell at given coordinate.
+    #
+    # coord - An X,Y coordinate passed as an Array.
+    #
+    # Example
+    #      -------------
+    #    0 | 1 | 2 | 3 |
+    #      -------------
+    #    1 | A | B | C |
+    #      -------------
+    #    2 | X | Y | Z |
+    #      -------------
+    #      - 0 - 1 - 2 -
+    #
+    #
+    #   find_cell([1,0])
+    #     #=> Y
+    #
+    #   find_cell([2,2])
+    #     #=> 3
+    #
+    # find_cell([0,1])
+    #     #=> A
+    #
+    def find_cell(coord)
+      board.find {|cell| cell.coord.position == coord}
+    end
+
+    # Public: Check if the gameboard is full relative to the default cell.
+    #
+    #
+    #
+    # Example
+    #
+    #    Board.new(height: 3, width: 3)
+    #      -------------
+    #    0 | 1 | 2 | 3 |
+    #      -------------
+    #    1 | A | B | C |
+    #      -------------
+    #    2 | X | Y | Z |
+    #      -------------
+    #      - 0 - 1 - 2 -
+    #
+    #   full?
+    #     #=> true
+    #
+    #   Board.new(height: 3, width: 3, cells: "Y")
+    #
+    #   full?
+    #     #=> false (because our default piece is now "Y")
+    #
+    def full?
+      board.none? { |cell| cell.value == cells  }
```

Comment:
This is where the vagueness of the *cells* property comes back to cause trouble. Is the board full if none of the cell values are equal to the board's default cell value? What if a user sets the default value to be a piece of some sort? Imagine a game where the board starts as full and is cleared over time. Then the board would be full, but this method would not return true. If *cells* in fact means default_empty_state, then that should be signified in the name.